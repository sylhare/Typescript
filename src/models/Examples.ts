export const workTripHoliday = [{ name: '🧳' }, { name: '✈️' }, { name: '🌴' }];

export const people = [
  {
    person: '👩‍🔬',
    type: 'scientist',
    tools: ['🔬', '🧪', '⚗️']
  },
  {
    person: '👩‍🔧',
    type: 'worker',
    tools: ['🔨', '🔧', '🔩', '⚙️']
  },
  {
    person: '👩‍🍳',
    type: 'cook',
    tools: ['🍳']
  }
];
