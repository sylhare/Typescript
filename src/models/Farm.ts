export const farm = {
  pig: {
    type: 'animal',
    img: '🐖'
  },
  cow: {
    type: 'animal',
    img: '🐄'
  },
  tractor: {
    type: 'building',
    img: '🚜'
  },
  farmer: {
    type: 'human',
    img: '👩‍🌾'
  }
};

export type FruitBasket = { apple: string, banana: string, kiwi: string };
export const fruitBasket = {
  apple: '🍎',
  banana: '🍌',
  kiwi: '🥝',
};

export type Fruit = { name: string, emoji: string };
export const listOfFruits = [
  { name: 'apple', emoji: '🍎' },
  { name: 'banana', emoji: '🍌' },
  { name: 'kiwi', emoji: '🥝' },
];
