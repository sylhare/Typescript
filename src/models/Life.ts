export const workTripHoliday = [{ name: '🧳' }, { name: '✈️' }, { name: '🌴' }];

export const people = [
  {
    person: '👩‍🔬',
    type: 'scientist',
    tools: ['🔬', '🧪', '⚗️']
  },
  {
    person: '👩‍🔧',
    type: 'worker',
    tools: ['🔨', '🔧', '🔩', '⚙️']
  },
  {
    person: '👩‍🍳',
    type: 'cook',
    tools: ['🍳']
  }
];

export const fruitBasket = {
  apple: '🍎',
  banana: '🍌',
  kiwi: '🥝',
};

export const listOfFruits = [
  { name: 'apple', emoji: '🍎' },
  { name: 'banana', emoji: '🍌' },
  { name: 'kiwi', emoji: '🥝' },
];
