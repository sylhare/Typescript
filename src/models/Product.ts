export class Product {
  id: number = 0;
  name: string = 'Product';

  constructor(product?: Partial<Product>) {
    Object.assign(this, product);
  }

  toString() {
    return 'This is ' + this.name;
  }
}

export function randomProduct() {
  const id = Math.floor(Math.random() * 100);
  return new Product({ id, name: `Product-${id}` });
}
