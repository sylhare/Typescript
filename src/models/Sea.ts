export type SeaCreature = { fish: string, deadly: boolean };

export const deadliestCreature: SeaCreature = { fish: '🦐', deadly: true };

export const sea: SeaCreature[] = [
  deadliestCreature,
  { fish: '🐡', deadly: false },
  { fish: '🐠', deadly: false },
  { fish: '🦈', deadly: false },
  { fish: '🦀', deadly: false },
];
