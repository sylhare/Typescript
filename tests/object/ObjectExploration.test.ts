import { farm, Fruit, fruitBasket, FruitBasket, listOfFruits } from '../../src/models/Farm';
import { people, workTripHoliday } from '../../src/models/Examples';

describe('Object exploration', () => {
  describe('Farm', () => {
    const entries = Object.entries(farm);

    it('can be viewed as entries', () => {
      expect(entries).toMatchSnapshot();
      expect(entries.length).toEqual(4);
      expect(entries[0].length).toEqual(2);
    });

    it('takes entries back to an object', () => {
      expect(Object.fromEntries(entries)).toEqual(farm);
    });

    it('groups by type', () => {
      const groupedFarm = entries.reduce((result: any, [key, value]) => ({
        ...result,
        [value.type]: [...(result[value.type] || []), ({ ...value, name: key })],
      }), {});
      expect(groupedFarm).toMatchSnapshot();
      expect(groupedFarm.animal.length).toEqual(2);
      expect(groupedFarm.human[0]).toMatchObject({
        ...farm.farmer,
        name: 'farmer',
      });
    });
  });

  describe('Fruit basket', () => {
    it('takes a fruit basket into a list of fruits', () =>
      expect(Object.entries(fruitBasket).map(([key, value]) => ({ name: key, emoji: value })))
        .toEqual(listOfFruits));

    it('takes a list of fruits into a fruit basket', () =>
      expect(Object.fromEntries(listOfFruits.map(it => [it.name, it.emoji])))
        .toEqual(fruitBasket));

    it('takes a TYPED fruit basket into a TYPED list of fruits', () =>
      expect(Object.entries(fruitBasket as FruitBasket).map(([key, value]) => ({ name: key, emoji: value })))
        .toEqual(listOfFruits as Fruit[]));

    it('takes a TYPED list of fruits into a TYPED fruit basket', () =>
      expect(Object.fromEntries((listOfFruits as Fruit[]).map(it => [it.name, it.emoji])))
        .toEqual(fruitBasket as FruitBasket));
  });

  describe('Other examples', () => {

    it('should be mapped to a name', () =>
      expect(workTripHoliday.map(it => it.name).join('-'))
        .toEqual('🧳-✈️-🌴'));

    it('life works in mysterious ways', () =>
      expect(people.map(it => ({ ...it, hobby: workTripHoliday })))
        .toMatchObject(people));
  });
});