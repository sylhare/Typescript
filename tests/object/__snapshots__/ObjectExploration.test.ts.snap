// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Object exploration Farm can be viewed as entries 1`] = `
[
  [
    "pig",
    {
      "img": "🐖",
      "type": "animal",
    },
  ],
  [
    "cow",
    {
      "img": "🐄",
      "type": "animal",
    },
  ],
  [
    "tractor",
    {
      "img": "🚜",
      "type": "building",
    },
  ],
  [
    "farmer",
    {
      "img": "👩‍🌾",
      "type": "human",
    },
  ],
]
`;

exports[`Object exploration Farm groups by type 1`] = `
{
  "animal": [
    {
      "img": "🐖",
      "name": "pig",
      "type": "animal",
    },
    {
      "img": "🐄",
      "name": "cow",
      "type": "animal",
    },
  ],
  "building": [
    {
      "img": "🚜",
      "name": "tractor",
      "type": "building",
    },
  ],
  "human": [
    {
      "img": "👩‍🌾",
      "name": "farmer",
      "type": "human",
    },
  ],
}
`;
